ykcapi/src/lib.rs

```diff
@@ -48,6 +52,99 @@ pub extern "C" fn __ykutil_get_llvmbc_section(res_addr: *mut *const c_void, res_
+        }
+        self.read_from_stack(id.try_into().unwrap())
+    }
+}
+
+/// Parses the stackmap and saved registers from the given address (i.e. the return address of the
+/// deoptimisation call).
+/// FIXME: Until we have the stopgap interpreter we simply print out the values we find.
+#[cfg(target_arch = "x86_64")]
+#[no_mangle]
+pub extern "C" fn yk_stopgap(
+    sm_addr: *const c_void,
+    sm_size: usize,
+    retaddr: usize,
+    rsp: *const c_void,
+) {
+    eprintln!("jit-state: stopgap");
+    // Restore saved registers from the stack.
+    let registers = Registers::from_ptr(rsp);
+
+    // Parse the stackmap.
+    let slice = unsafe { slice::from_raw_parts(sm_addr as *mut u8, sm_size) };
+    let map = StackMapParser::parse(slice).unwrap();
+    let locs = map.get(&retaddr.try_into().unwrap()).unwrap();
+
+    // Extract live values from the stackmap.
+    for l in locs {
+        match l {
+            SMLocation::Register(reg, size) => {
+                eprintln!("Register: {} ({} {})", registers.get(*reg), reg, size);
+            }
+            SMLocation::Direct(reg, off, size) => {
+                // When using `llvm.experimental.deoptimize` then direct locations should always be
+                // in relation to RSP.
+                assert_eq!(*reg, 7);
+                // We need to add 2 bytes to the value of RSP to factor in the return address and
+                // the pushed RBP value of the previous frame.
+                let addr = registers.get(*reg) + (*off as usize) + 16;
+                let v = match *size {
+                    1 => unsafe { ptr::read::<u8>(addr as *mut u8) as u64},
+                    2 => unsafe { ptr::read::<u16>(addr as *mut u16) as u64},
+                    4 => unsafe { ptr::read::<u32>(addr as *mut u32) as u64},
+                    8 => unsafe { ptr::read::<u64>(addr as *mut u64) as u64},
+                    _ => unreachable!(),
+                };
+                eprintln!("Direct: {} ({} {})", v, reg, off);
+            }
+            SMLocation::Indirect(reg, off, size) => {
+                let addr = registers.get(*reg) + (*off as usize) + 16;
+                let v = match *size {
+                    1 => unsafe { ptr::read::<u8>(addr as *mut u8) as u64},
+                    2 => unsafe { ptr::read::<u16>(addr as *mut u16) as u64},
+                    4 => unsafe { ptr::read::<u32>(addr as *mut u32) as u64},
+                    8 => unsafe { ptr::read::<u64>(addr as *mut u64) as u64},
+                    _ => unreachable!(),
+                };
+                eprintln!("Indirect: {} ({} {})", v, reg, off);
+            }
+            SMLocation::Constant(v) => {
+                eprintln!("Constant: {}", v);
+            }
+            SMLocation::LargeConstant(v) => {
+                eprintln!("Large constant: {}", v);
+            }
+        }
+    }
+    // FIXME: Initialise stopgap interpreter here.
+    process::exit(0);
```

Comment:
Wouldn't a `todo` be better? Or is this to ensure that tests can run?

Reply:
The latter.